### DISCUSSION TITLE
convert package to library

### DISCUSSION THREAD

**Discussion by prfpeste** [OpenModelica/OpenModelica#12673] [Ideas]
Hi there,
currently for whatever reason I fail with the convertPackageToLibrary skript. The intention is to convert my own library XXX.mo into a Modelica library with the typical folder structure (package.mo, package.order etc.).

Nevertheless, the commands in OM CLI:

loadFile("/path/file.mo")
convertPackageToLibrary("file","/path/to/file","1.0.0", true)

doesn't work. The library is load, but the conversion doesn't work.

The manual was here not very helpful for me. For this reason, and after searching on the web for a solution without any success, I think it would be great if the function could be implemented in OMEdit directly as an export to library button. I'm sure this would be of great use for several developers that don't use Dymola.

Best regards P.

**Comment by casella** [comment]
I have to admit I was not aware of this API function, and the [documentation ](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html#convertpackagetolibrary) doesn't really explain what it does.

@perost, @adrpo do you know what is API function good for?

**Reply by perost** [reply]
I don't know why it doesn't show up in the documentation, but the function does have a documentation annotation:
> Runs the conversion script for a library on a selected package.

In other words, it converts a package to a specific version of a library using the library's conversion script, for example if you have a package that you want to convert to MSL 4.0.0.

**Reply by prfpeste** [reply]
Ok, this doesn't sound like what I inteded to do with it. But is there an option to automatically convert a .mo file into a Modelica library?
If not, this would be in my mind a great feature.

**Reply by casella** [reply]
OK, this makes sense. The documentation is of course essential, otherwise users such as @prfpeste can get badly mislead about the purpose of the function 😅 

The documentation annotation is very terse. What about 

> Runs the appropriate conversion script to upgrade a loaded package packageToConvert that uses an old version of library to the new version libraryVersion of the same library. For example, convertPackageToLibrary(MyPackage, Modelica, "4.0.0") converts MyPackage to use version 4.0.0 of the Modelica standard library, instead of the currently used old version, e.g. 3.2.3.

**Reply by casella** [reply]
@perost see my [comment in #11033](https://github.com/OpenModelica/OpenModelica/issues/11033#issuecomment-2200789732). Thanks!

**Reply by casella** [reply]
> But is there an option to automatically convert a .mo file into a Modelica library?

A Modelica library is a Modelica package. According to [Section 13.4](https://specification.modelica.org/maint/3.6/packages.html#file-system-mapping-of-package-class) of the specification, this can be mapped into a range of different ways, from a single .mo file with all the nested packages in it, to a fully expanded file system structure where each class is saved in a single file. So, you are not converting a .mo file into a package, you need to convert a certain file system representation of a given package into another file system representation of the same package.

@prfpeste I'm not sure we already have such a feature, but I guess this could be implemented easily in OMEdit. Please open a proper ticket for that. Thanks!
